# <center>Pandas HR database Query [24 exercises with solution]</center>

![HR Database Schema](https://www.w3resource.com/w3r_images/database-model-hr-new.gif)

In [1]:
# Reading the CSV files:

import pandas as pd
pd.set_option('display.max_rows', 500)
pd.set_option('display.max_columns', 500)

employees = pd.read_csv(r"EMPLOYEES.csv")
departments = pd.read_csv(r"DEPARTMENTS.csv")
job_history = pd.read_csv(r"JOB_HISTORY.csv")
jobs = pd.read_csv(r"JOBS.csv")
countries = pd.read_csv(r"COUNTRIES.csv")
regions = pd.read_csv(r"REGIONS.csv")
locations = pd.read_csv(r"LOCATIONS.csv")

## Exploring the files:

In [7]:
regions.head()

REGION_ID             REGION_NAME
0          1                  Europe
1          2                Americas
2          3                    Asia
3          4  Middle East and Africa

In [8]:
countries.head()

country_id country_name  region_id
0         AR    Argentina          2
1         AU    Australia          3
2         BE      Belgium          1
3         BR       Brazil          2
4         CA       Canada          2

In [9]:
locations.head()

location_id           street_address postal_code       city  \
0         1000     1297 Via Cola di Rie         989       Roma   
1         1100  93091 Calle della Testa       10934     Venice   
2         1200         2017 Shinjuku-ku        1689      Tokyo   
3         1300          9450 Kamiya-cho        6823  Hiroshima   
4         1400      2014 Jabberwocky Rd       26192  Southlake   

     state_province country_id  
0               NaN         IT  
1               NaN         IT  
2  Tokyo Prefecture         JP  
3               NaN         JP  
4             Texas         US

In [10]:
employees.head()

employ_id first_name last_name     email  phone_number hire_date   job_id  \
0        100     Steven      King     SKING  515.123.4567  17-06-87  AD_PRES   
1        101      Neena   Kochhar  NKOCHHAR  515.123.4568  18-06-87    AD_VP   
2        102        Lex   De Haan   LDEHAAN  515.123.4569  19-06-87    AD_VP   
3        103  Alexander    Hunold   AHUNOLD  590.423.4567  20-06-87  IT_PROG   
4        104      Bruce     Ernst    BERNST  590.423.4568  21-06-87  IT_PROG   

   salary  commission_pct  manager_id  department_id  
0   24000             NaN         NaN           90.0  
1   17000             NaN       100.0           90.0  
2   17000             NaN       100.0           90.0  
3    9000             NaN       102.0           60.0  
4    6000             NaN       103.0           60.0

In [11]:
departments.head()

department_id  department_name  manager_id  location_id
0             10   Administration       200.0         1700
1             20        Marketing       201.0         1800
2             30       Purchasing       114.0         1700
3             40  Human Resources       203.0         2400
4             50         Shipping       121.0         1500

In [12]:
jobs.head()

job_id                      job_title  min_salary  max_salary
0     AD_PRES                      President       20000       40000
1       AD_VP  Administration Vice President       15000       30000
2     AD_ASST       Administration Assistant        3000        6000
3      FI_MGR                Finance Manager        8200       16000
4  FI_ACCOUNT                     Accountant        4200        9000

In [4]:
job_history.head()

employee_id  start_date   end_date`      job_id  department_id
0          102  1993-01-13  1998-07-24     IT_PROG             60
1          101  1989-09-21  1993-10-27  AC_ACCOUNT            110
2          101  1993-10-28  1997-03-15      AC_MGR            110
3          201  1996-02-17  1999-12-19      MK_REP             20
4          114  1998-03-24  1999-12-31    ST_CLERK             50

## Solving Exercises Problems:

### 1. Write a Pandas program to display all the records of REGIONS file.

### 2. Write a Pandas program to display all the location id from locations file.

### 3. Write a Pandas program to extract first 7 records from employees file.

### 4. Write a Pandas program to select distinct department id from employees file. 

### 5. Write a Pandas program to display the first and last name, and department number for all employees whose last name is "McEwen".

### 6. Write a Pandas program to display the first, last name, salary and department number for those employees whose first name starts with the letter 'S'.

### 7. Write a Pandas program to display the first, last name, salary and department number for those employees whose first name does not contain the letter 'M'.

### 8. Write a Pandas program to display the first name, last name, salary and department number in ascending order by department number. 

### 9. Write a Pandas program to display the first name, last name, salary and department number in descending order by first name.

### 10. Write a Pandas program to display the first name, last name, salary and manger id where manager ids are null.

### 11. Write a Pandas program to display the first name, last name, salary and manger id where manager ids are not null.

### 12. Write a Pandas program to create and display a boolean series, where True for not null and False for null values or missing values in state_province column of locations file.

### 13. Write a Pandas program to create a boolean series selecting rows with one or more nulls from locations file.

### 14. Write a Pandas program to count the NaN values of all the columns of locations file. 

### 15. Write a Pandas program to display the first name, last name, salary and department number for those employees whose first name ends with the letter 'm'.

### 16. Write a Pandas program to display the first name, last name, salary and department number for those employees whose first name ends with the letter 'd' or 'n' or 's' and also arrange the result in descending order by department id.

### 17. Write a Pandas program to display the first name, last name, salary and department number for employees who works either in department 70 or 90.

### 18. Write a Pandas program to display the first name, last name, salary and department number for those employees whose managers are hold the ID 120, 103 or 145.

### 19. Write a Pandas program to display the first, last name, salary and department number for those employees who holds a letter n as a 3rd character in their first name.

### 20. Write a Pandas program to display the first name, job id, salary and department for those employees not working in the departments 50,30 and 80.

### 21. Write a Pandas program to display the ID for those employees who did two or more jobs in the past.

### 22. Write a Pandas program to calculate minimum, maximum and mean salary from employees file.

### 23. Write a Pandas program to display the details of jobs in descending sequence on job title.

### 24. Write a Pandas program to display the first and last name and date of joining of the employees who is either Sales Representative or Sales Man.

# <center>The END</center>

#### Reference: [https://www.w3resource.com/python-exercises/pandas/sql/index.php](https://www.w3resource.com/python-exercises/pandas/sql/index.php)